rfcs/0144-versioned-flake-references.md

```diff
@@ -0,0 +1,118 @@
+library authors to introduce breaking changes in a communicatable way that will
+not break their downstream dependents's code.
+
+# Detailed design
+[design]: #detailed-design
+
+## Syntax
+Version placeholders will be in `{}`, with a list of version requirements
+separated by `,` (commas), e.g. `{^1.0,<3}` will expand to a version that
+specifies the version requirements `^1.0` and `<3`
+
+Version requirements consists of a comparator and version. The version does not
+have to be a valid version defiined by the versioning scheme, `1` and `1.0` are
+also valid versions for version requirements. The comparator has to be one of
+the following options:
+- `^` compatible (as defined by the versioning scheme)
+- `<` less than
+- `<=` less than or equal to
+- `>` greater than
+- `>=` greater than or equal to
+- `=` equal to
+
+A version placeholder with no version requirements will match all valid
+versions defined by the versioning scheme.
+
+## Versioning scheme
+The versions will follow [semantic versioning] (semver). The tags have to
+follow semver, `1.0` and `1.0.0.0` will not match the placeholder `{=1.0}` as
+it is not a valid version defined by semver. Build identifiers will not be
+taken into consideration when calculating version requirements, and the latest
+tag will be selected if multiple tags have the same precedence defined by
+semver.
+
+## Upgrading
+A new Nix command, `nix flake upgrade`, will edit `flake.nix` and upgrade all
+the version placeholders in the inputs. A new flag, `--upgrade-input` will
+upgrade only the specified flake input. Only version placeholders that contain
+exactly one version requirement which the comparator is either `^` or `=` will
+be upgraded. All other version placeholders will be left unchanged.
+
+# Examples and Interactions
+[examples-and-interactions]: #examples-and-interactions
+
+Here are the git tags of a hypothetical Nix flake "github:foo/bar":
+- v0.1.0
+- v0.1.1
+- v0.2.0
+- v1.0.0
+- v1.0.1
+- v1.1.0
+- v2.0.0
+- unrelated
+
+Flake reference | The tag it points to | The `upgrade`ed flake reference
+-|-|-
+`github:foo/bar/v{^1}` | `v1.1.0` | `github:foo/bar/v{^2}`
+`github:foo/bar/v{=1.0}` | `v1.0.1` | `github:foo/bar/v{=2.0}`
+`github:foo/bar/v{^1.0}` | `v1.1.0` | `github:foo/bar/v{^2.0}`
+`github:foo/bar/v{=1.0.0}` | `v1.0.0` | `github:foo/bar/v{=2.0.0}`
+`github:foo/bar/v{^1.0.0}` | `v1.1.0` | `github:foo/bar/v{^2.0.0}`
+`github:foo/bar/v{^0.1.0}` | `v0.1.1` | `github:foo/bar/v{^2.0.0}`
+`github:foo/bar/v{<2}` | `v1.1.0` | `github:foo/bar/v{<2}` (no change)
+`github:foo/bar/v{>1.0}` | `v2.0.0` | `github:foo/bar/v{>1.0}` (no change)
+`github:foo/bar/v{^1,<1.1}` | `v1.0.1` | `github:foo/bar/v{^1,<1.1}` (no change)
+`github:foo/bar/v{}` | `v2.0.0` | `github:foo/bar/v{}` (no change)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- This is impossible with some types of flake references, such as `path`s and
+  `git+` URLs, which will make the flake interface less consistent.
```

Comment:
Why is it impossible with `git+` URLs? Seems like the prerequisite is an input type that supports enumerating a set of versions (e.g. by getting the tags of a repo).